### ISSUE TITLE
bug: color inaccuracy

### ISSUE THREAD

**Issue by ThaUnknown** [libass/JavascriptSubtitlesOctopus#152]
I remember there being talk about some cases where JSSO has color inaccuracies in the bitmap decode calculation and I reproduced this on `[SakuraCircle] Acchi Kocchi - 09 (BD 1920x1080 HEVC FLAC) [CB0433CA].mkv` where `fd91a0` is calculated as `f484a1` @ 15:51

**Comment by ThaUnknown** [comment]
this is because the .ass subtitles included `YCbCr Matrix: TV.601` in their script info. does libass compensate the matrix? or does it not? if it does is there a way to turn that off?

**Comment by ThaUnknown** [comment]
full file: [4Tsumiki.ass.txt](https://github.com/libass/JavascriptSubtitlesOctopus/files/9381981/4Tsumiki.ass.txt)

**Comment by TheOneric** [comment]
JSO needs to know the video colourspace to accurately deal with most ASS colours. Read the docs:
https://github.com/libass/libass/blob/b890df23eb8931728acdf240fd5d29e53401b4f2/libass/ass_types.h#L152-L217

> this is because the .ass subtitles included `YCbCr Matrix: TV.601` in their script info.

Not putting any `YCbCr` header and putting an explicit `YCbCr Matrix: TV.601` are identical in all compatible ASS renderers.

**Comment by TheOneric** [comment]
> so then is this not the issue causing this?

It most likely *is* the cause.

**Comment by ThaUnknown** [comment]
> > so then is this not the issue causing this?
> 
> It most likely _is_ the cause.

hmmmmm if TV.601 is the default value then I don't really know what else to do here... JSSO simply cant handle this case?

**Comment by ThaUnknown** [comment]
@TheOneric is the only way to fix this problem is to know the video metadata, aka color space? I've been trying to read up on this, but haven't gotten much information as to implementation details. I've found a way to extract colorspace metadata from videos in browser, if that is going to be of any help?

**Comment by TheOneric** [comment]
> is the only way to fix this problem is to know the video metadata, aka color space?

Yes, this is a necessary part of the solution.

> I've been trying to read up on this, but haven't gotten much information as to implementation details.

The link to the related libass docs in my first reply here already contains the full details. *(Since there were some rewordings since, I also just updated the link to point at the newest version.)*

**Comment by ThaUnknown** [comment]
> > is the only way to fix this problem is to know the video metadata, aka color space?
> 
> Yes, this is a necessary part of the solution.
> 
> > I've been trying to read up on this, but haven't gotten much information as to implementation details.
> 
> The link to the related libass docs in my first reply here already contains the full details. _(As there were some rewordings since, I also just updated the link to point at the newest version.)_

I could probably fix this, but I'm not certain what actually would need to be done in this case, could simply changing what color space is given to libass itself in the subtitle file be a potential fix? or does it need to be done after libass spits out its images

**Comment by ThaUnknown** [comment]
seems like we could "borrow" some of the MPV code for this https://github.dev/mpv-player/mpv/blob/59fc8eecbc08d22368451b31fcf799aec557ab0d/sub/sd_ass.c

**Comment by Yay295** [comment]
Looks like it's this function specifically:

https://github.com/mpv-player/mpv/blob/59fc8eecbc08d22368451b31fcf799aec557ab0d/sub/sd_ass.c#L853-L958

**Comment by ThaUnknown** [comment]
yeah, but there's a lot of other black magic happening there that I honestly can't wrap my head around and I'm not sure what a lot of it does :/

**Comment by Yay295** [comment]
More stuff here too: https://github.com/mpv-player/mpv/blob/f02ab2aabb630d6618dd99f63af8014c10d79562/video/csputils.c

**Comment by ThaUnknown** [comment]
@TheOneric if you ever want to come back to this, I can see 2 ways of doing this, the same way that MPV does it, so pretty much steal their code for this, or use SVG filters, TLDR a [canvas context can have a filter](https://devdocs.io/dom/canvasrenderingcontext2d/filter) which can be an URL, which can be an SVG filter, those can do [color matrix conversions](https://developer.mozilla.org/en-US/docs/Web/SVG/Element/feColorMatrix) but I can't find any reliable info on how to do the math for the RGB conversion for this, but the matrixes can do:
```
     R G B A W
R' | 1 0 0 0 0 |
G' | 0 1 0 0 0 |
B' | 0 0 1 0 0 |
A' | 0 0 0 1 0 |
```
where W is hard offset
